src/librustc_typeck/check/mod.rs

```diff
@@ -3853,6 +3853,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                   self.diverges.set(cond_diverging);
```

Comment:
Ah, this is the bit I was missing before (which ignores the divergence of the body). Very good.